### PR TITLE
Fix bigarray config (#5494)

### PR DESCRIPTION
**Pull request by moyodiallo**
This is a tentative fix for #5494, bigarray issue on OCaml 5.00
I explored 2 solutions and I chose the second one.

- It's to avoid generating bigarray in Meta files like mentioned in the issue #5494. But with the generation of Meta files, bigarray has to be resolved before and that raises an exception.
- The second is to avoid resolving bigarray, that also doesn't generate it in the Meta files. 

I'm not sure this is the right way to go about this problem, let me know if anyone thought of a better solution and I can make adjustments to this PR.

cc @kit-ty-kate

### CODE REVIEW

**Comment by dra27** [comment]
Thanks for this! Comparing opam-repository for current trunk OCaml (with [opam-alpha-repository](https://github.com/kit-ty-kate/opam-alpha-repository) enabled), this unblocks 56 packages and reveals 7 new failing packages - several of which look they have pertinent errors.

**Comment by dra27** [comment]
Thanks for this! Comparing opam-repository for current trunk OCaml (with [opam-alpha-repository](https://github.com/kit-ty-kate/opam-alpha-repository) enabled), this unblocks 56 packages and reveals 7 new failing packages - several of which look they have pertinent errors.

File: src/dune_rules/lib.ml [hunk 1]
```diff
@@ -1435,14 +1435,26 @@ end = struct
...
           | Direct (loc, name) ->
-            let+ lib = resolve_dep db (loc, name) ~private_deps in
-            let acc_res =
-              let open Resolve.O in
-              let+ lib = lib
-              and+ acc_res = acc_res in
-              lib :: acc_res
-            in
-            (acc_res, acc_selects, acc_re_exports)
+            let ocaml_version = db.lib_config.ocaml_version in
```
Comment (emillon) [inline src/dune_rules/lib.ml:1438]: Not an expert with lib rules, but this feels very early in the pipeline to filter out `bigarray`. For example it will ignore `bigarray` if it's under a `(select)`. I think that it should be done later, after everything has been merged together.

File: src/dune_rules/lib.ml [hunk 2]
```diff
@@ -1435,14 +1435,26 @@ end = struct
...
+            if String.equal "bigarray" (Lib_name.to_string name)
+              && not (Ocaml_version.has_bigarray_library ocaml_version)
+            then
+              begin
+                User_warning.emit ~loc
+                  [ Pp.textf
+                      "Already in the Standard Library of OCaml %s"
```
Comment (emillon) [inline src/dune_rules/lib.ml:1445]: I think that this warning might not be actionable enough: if a project builds on 4.04+ and 5, we don't want it to warn on all the versions. Dune does not have access to the minimum supported ocaml version so I don't think there's a good way to solve that in dune.

File: src/dune_rules/lib.ml (hunk 1 above)
Comment (moyodiallo) [reply src/dune_rules/lib.ml:1438]: This part of the source code is executed Lazily. Almost everything in dune is merged together lazily. Waiting later is like waiting when the package will be read (in dune system resolving the package) to be put in META files like I said https://github.com/ocaml/dune/pull/5526#issue-1175246595, at the reading an error will be raised. Avoiding the error in this function in order to skip generating the package in META files, implies introducing a specific no-op result type and I think that can affect some part of the source code which is not necessary.

I did some inspection, `(select)`  is not used with bigarray right now in opam repository, but `(libraries)` yes:
- `opam grep --regex="(libraries.*bigarray" | wc -l` give 371 package
- `opam grep --regex="(select.*bigarray" | wc -l` give 0 package

File: src/dune_rules/lib.ml (hunk 2 above)
Comment (moyodiallo) [reply src/dune_rules/lib.ml:1445]: I think the aim of this PR, is to fix dune for OCaml 5.00 without breaking the old state of the ecosystem. I can't see the interest of taking account OCaml < 4.07.
@dra27 correct me if I'm wrong.

Comment (emillon) [reply src/dune_rules/lib.ml:1445]: let me rephrase what I mean. the difference I'm highlighting is between:
1. add a warning if the presence of `bigarray` in `(libraries)` is superfluous for the project because it requires ocaml 4.07 or later (point 2 in #5494)
2. add a warning if the presence of `bigarray` in `(libraries)` is found but discarded because of the ocaml version we're compiling with (what this PR does if I understand it correctly)

We want every warning to be actionnable: there should be a way to fix the project so that it builds cleanly with all its supported configurations.

In the case 1 above, the warning has a clear action: remove `bigarray` in this `(libraries)` field. This is not going to break anything and will make the warning go away. (but this strategy is difficult to implement within dune itself and I think would fit better in a lint tool)

In the case number 2, `bigarray` cannot be removed because doing so will break ocaml 4 builds. So this means that a project that supports both ocaml 4 and 5 is bound to have warnings in the ocaml 5 builds.

I think that the question we need to find an answer for is, how does the `(libraries)` field look like in a project that uses the `Bigarray` module and supports both OCaml 4 and OCaml 5?

> I can't see the interest of taking account OCaml < 4.07.

Not sure what you mean by this, but dune continues to support building projects with OCaml 4.02+.

Comment (moyodiallo) [reply src/dune_rules/lib.ml:1445]: > ''I can't see the interest of taking account OCaml < 4.07.''
> Not sure what you mean by this, but dune continues to support building projects with OCaml 4.02+.

What I meant was that in the ecosystem it's well known OCaml < 4.07 has not `bigarray` library.

I get what your are highlighting. Another suggestion: Is it a good idea to avoid the warning (silent mode) ? 
In that case a project with OCaml 4.07+ config won't be confusing because resolving a package take account the compiler version at that moment if I'm not mistaken.

> In the case number 2, bigarray cannot be removed because doing so will break ocaml 4 builds.

Do you mean during the parsing of dune files or when we're compiling depends on OCaml version?
Did dune generate the same META files for different ocaml version like a project with OCaml 4 and 5 config ? (I'm not a expert in dune at all)

File: src/dune_rules/lib.ml (hunk 1 above)
Comment (dra27) [reply src/dune_rules/lib.ml:1438]: I think this is the right place, but the other cases need doing. The `Re_export` case is the same as the `Direct` case, so I expect the code here can be hoisted and shared a little (i.e. `(re_export bigarray)` should be a no-op in the same way as `bigarray`).

`select` should be implemented - `!bigarray` should cause a select filter _to fail_ (because bigarray is always there) so a scan of `forbidden` is needed to check for it. For the positive case, `bigarray` should be filtered out of `required`. I quickly tested on 5.0 only with:

```diff
--- a/src/dune_rules/lib.ml
+++ b/src/dune_rules/lib.ml
@@ -1388,11 +1388,15 @@ end = struct
           Memo.Build.List.find_map choices
             ~f:(fun { required; forbidden; file } ->
               let forbidden = Lib_name.Set.to_list forbidden in
+              if List.exists ~f:(fun lib -> String.equal (Lib_name.to_string lib) "bigarray") forbidden
+              then Memo.Build.return None
+              else
               let* exists =
                 Memo.Build.List.exists forbidden ~f:(available_internal db)
               in
               if exists then Memo.Build.return None
               else
+              let required = Lib_name.Set.filter ~f:(fun lib -> not(String.equal (Lib_name.to_string lib) "bigarray")) required in
                 Resolve.Build.peek
                   (let deps =
                      Lib_name.Set.fold required ~init:[] ~f:(fun x acc ->
```
but better code will be available, and it needs the version check. However, those two additions cause this `dune` file to be evaluated correctly in 5.0:

```
(executable
  (name c)
  (libraries (select c.ml from (!bigarray -> c.nobigarray.ml) (bigarray -> c.bigarray.ml) ( -> c.dummy.ml))))
```

without the check on `forbidden`, `c.nobigarray.ml` would be selected and obviously without the filter on `required` the `c.bigarray.ml` would never be selected.

File: src/dune_rules/lib.ml (hunk 2 above)
Comment (dra27) [reply src/dune_rules/lib.ml]: @moyodiallo - that's not quite right, the bigarray library is present in 4.07 and earlier, too - what changed is that it was added to the Standard Library. The point of this change is to allow a user to specify `(libraries bigarray)` and have that work on OCaml 4.02 and 5.0 - i.e. it _adds_ the bigarray library on 4.02 (you must link with `bigarray.cm[x]a`) but is a no-op on 5.0, where `Bigarray` will be used from the Standard Library.

**Comment by dra27** [comment]
As it happens, I started reviewing and poking this today, too, @nojb! I think the other two cases need (`re_export` and `select`) and then it's good to go. It can indeed be merged without the warning.

FWIW, it should be possible to extend this behaviour back to 4.08 - IIRC since 4.08 it has not been necessary ever to use the legacy bigarray library, but it would be better to do that in a separate PR (if at all), as well.

File: src/dune_rules/lib.ml (hunk 2 above)
Comment (emillon) [reply src/dune_rules/lib.ml]: Thanks @dra27 ! I think that the `Ocaml_version.has_bigarray_library` function can be renamed to reflect this - `stdlib_includes_bigarray` (with opposite meaning) seems more precise.

File: src/dune_rules/lib.ml (hunk 1 above)
Comment (emillon) [reply src/dune_rules/lib.ml:1438]: Great. Can you include this as a new test case?

File: src/dune_rules/lib.ml (hunk 2 above)
Comment (moyodiallo) [reply src/dune_rules/lib.ml]: > Thanks @dra27 ! I think that the `Ocaml_version.has_bigarray_library` function can be renamed to reflect this - `stdlib_includes_bigarray` (with opposite meaning) seems more precise.

@emillon  in the current PR ?

Comment (emillon) [reply src/dune_rules/lib.ml]: Hmm there's already a `stdlib_includes_bigarray` in `Ocaml_version`, with a different meaning. So let's forget about this for now.

File: src/dune_rules/lib.ml (hunk 1 above)
Comment (moyodiallo) [reply src/dune_rules/lib.ml:1438]: @emillon regarding the new test case, I'm not familiar with the testsuites(not look in to it before).

**Comment by moyodiallo** [comment]
I think it's ready for last review then to be merged.

File: test/blackbox-tests/test-cases/bigarray-select.t [hunk 3]
```diff
@@ -0,0 +1,8 @@
+  $ echo "(lang dune 3.1)" > dune-project
+  $ cat >dune <<EOF
+  > (executable
+  >   (name c)
+  >   (libraries (select c.ml from (!bigarray -> c.nobigarray.ml) (bigarray -> c.bigarray.ml) ( -> c.dummy.ml))))
+  > EOF
+  $ touch c.bigarray.ml  c.dummy.ml  c.nobigarray.ml
+  $ dune build
```
Comment (kit-ty-kate) [inline test/blackbox-tests/test-cases/bigarray-select.t:8]: I’m not entirely sure what this test case is supposed to show. Unless I’m missing something, this would succeed before as well

Comment (moyodiallo) [reply test/blackbox-tests/test-cases/bigarray-select.t:8]: You're right, The test supposed to verify if `c.bigarray.ml` or `c.nobigarray.ml` appears in `_build/default/c`. If found then the test fail. I'm gonna fix it.

Comment (moyodiallo) [reply test/blackbox-tests/test-cases/bigarray-select.t:8]: I mean for OCaml 5.00 version of course.

**Comment by moyodiallo** [comment]
> `+              let required = Lib_name.Set.filter ~f:(fun lib -> not(String.equal (Lib_name.to_string lib) "bigarray")) required in`
Hi @dra27, I notice we could end-up require and forbidden empty so that switch quietly in the case "(_ -> ...)".

File: src/dune_rules/lib.ml [hunk 4]
```diff
@@ -1421,6 +1435,9 @@ end = struct
           let open Memo.Build.O in
           match (dep : Lib_dep.t) with
           | Re_export (loc, name) ->
+            if is_bigarray name && not bigarray_in_std_libraries
```
Comment (emillon) [inline src/dune_rules/lib.ml:1438]: Some parts (like the end of the `get` function) are a bit difficult to follow because the bigarray check is intertwined with the rest of the logic. I'll refactor that part to see if we can extract the bigarray logic in a more precise way.

**Review by emillon — COMMENTED** [review]
OK, let's get that one past the finish line. I'd like to make sure we can keep this logic not to spread around and we should be able to fix that for the next version of dune.

File: src/dune_rules/lib.ml [hunk 5]
```diff
@@ -1381,27 +1381,41 @@ end = struct
...
       let+ res, src_fn =
         let+ select =
           Memo.Build.List.find_map choices
             ~f:(fun { required; forbidden; file } ->
               let forbidden = Lib_name.Set.to_list forbidden in
-              let* exists =
-                Memo.Build.List.exists forbidden ~f:(available_internal db)
-              in
-              if exists then Memo.Build.return None
+              if List.exists ~f:(fun lib -> is_bigarray lib && not bigarray_in_std_libraries) forbidden
```
[the remark this answers is not in the document]
Comment (emillon) [reply src/dune_rules/lib.ml:1394]: got it - `forbidden` here does not refer to `(forbidden_libs)` but to `! lib`

[the remark this answers is not in the document]
Comment (moyodiallo) [reply src/dune_rules/lib.ml]: yeah and the goal is to filter bigarray against OCaml 5.00,  we evaluate only the old version. I think the condition is incomplete `is_bigarray lib && not (bigarray_in_std_libraries) && (available_internal db lib)`. 
lines 1401 - 1408 have to back on the old commit, I did it forgetting the goal is to filter.

[the remark this answers is not in the document]
Comment (moyodiallo) [reply src/dune_rules/lib.ml]: To clarify little bit this section of the code:

```(executable
  (name c)
  (libraries (select c.ml from (!bigarray -> c.nobigarray.ml) (bigarray -> c.bigarray.ml) ( -> c.dummy.ml))))
```

This example of @dra27  is decomposed in 3 choices:
1. `{require = []; forbidden[bigarray]; file="c.nobigarry.ml"}`
2. `{require = [bigarray]; forbidden[]; file="c.bigarry.ml"}`
3. `{require=[]; forbidden=[]; file "dummy"}`
---
All choices are evaluated and a choice's assumption is that a package in forbidden could not be in require.
The current flow of code start evaluating forbidden libraries before required ones of each choice.
When a forbidden library is found we skip the choice (normal action). 

Case 1 correspond to this condition.

File: test/blackbox-tests/test-cases/bigarray-skip/run.t [hunk 6]
```diff
@@ -0,0 +1,19 @@
+Bigarray library is included in the Standard libraries since OCaml 5.00.0.
```
[the remark this answers is not in the document]
Comment (moyodiallo) [reply test/blackbox-tests/test-cases/bigarray-skip/run.t:1]: The title is misleading, I consider the test for the old version too(including 4.14) without breaking the test.
This test gonna change little bit anyway.

[the remark this answers is not in the document]
Comment (dra27) [reply test/blackbox-tests/test-cases/bigarray-skip/run.t:1]: [bigstring](https://opam.ocaml.org/packages/bigstring/) is a perfect test for this!

**Comment by emillon** [comment]
I pulled the code in a way that adds a new "rewrite" step before the rest of the processing. This ensures that this PR is a no-op for OCaml 4.x. For the part that rewrites the lib dep, we create a new lib dep that is interpreted as before. This is done with the same rules we would use if we were upgrading a project to OCaml 5+.
The rules I used based on @moyodiallo's code are the following. In technical terms "rewriting" a list is a `filter_map` operation since we can drop or modify items.

- for direct dependendencies and reexports: drop them if the library is `bigarray`, or copy as is otherwise
- for each `select` form, rewrite into a `select` form. the list of choices is rewritten using the following rules:
  - if a choice contains `!bigarray` (negative / "forbidden"), remove it completely.
  - remove `bigarray` (positive / "required") from the list.
  - rationale: evaluating choices is like evaluating boolean formulas and we're making a partial evaluation. `!bigarray` is `False`, so `!bigarray & ...` evaluates to `False`. Since choices are evaluated sequentially, this is equivalent to removing the choice. And `bigarray` is `True`, which is neutral for `&` so it's equivalent to removing it from the choice.

There's a mildly interesting corner case which happens when no choices are left after rewriting. This happens with `(select a from (!bigarray -> b))`. The docs say that it's an error when no choice matches, so we should rewrite to a form that creates an error. We should need to be careful the select form with no choices if there's no concrete syntax for it, but fortunately (and a bit surprisingly), the following is already accepted and supported: `(select a from)`. So there's no special case to handle.

**Comment by dra27** [comment]
I rebased and propose https://github.com/ocaml/dune/pull/5526/commits/b2958dbe5442148e775c91c32913120cc73f1dba to update the test case - this runs three different programs which exercise the three cases in the code. The test is expected to run in the same way on any OCaml version - i.e. when Dune starts testing OCaml 5.0 in CI then that test should remain unchanged (I ran it locally with OCaml 5, of course).

**Comment by nojb** [comment]
> I rebased and propose [b2958db](https://github.com/ocaml/dune/commit/b2958dbe5442148e775c91c32913120cc73f1dba) to update the test case - this runs three different programs which exercise the three cases in the code. The test is expected to run in the same way on any OCaml version - i.e. when Dune starts testing OCaml 5.0 in CI then that test should remain unchanged (I ran it locally with OCaml 5, of course).

In https://github.com/ocaml/ocaml/pull/11007#discussion_r889495527 I suggested shipping `META.bigarray` in the compiler (I had forgotten about this PR), which seems to be at odds with the approach taken here for OCaml >= 5 (right?). To be honest, at this point I can't tell anymore which one is the "good" approach :) Anyway, it would be great if someone who has clarity about the issue were to add a comment (either for or against) in the discussion I linked. Thanks!

**Comment by dra27** [comment]
I'm just paging in the OCaml PR, but we definitely _shouldn't_ have a `META` file `bigarray`, because there is no bigarray package anymore (this _might_ get extended to bytes, but that's a separate question). (https://github.com/ocaml/dune/issues/5494, from a while ago now, has the fuller detail!)

**Comment by nojb** [comment]
> I'm just paging in the OCaml PR, but we definitely _shouldn't_ have a `META` file `bigarray`, because there is no bigarray package anymore (this _might_ get extended to bytes, but that's a separate question). (#5494, from a while ago now, has the fuller detail!)

Right, I'll make a note upstream, thanks for the clarifying.

**Comment by dra27** [comment]
> Could it be necessary to test which file is emitted from the case 3 (`select`) according to our fix, to give this test more sense.

It does - each of the files prints a different message, so if the wrong file is selected the cram test will fail because of different output.

**Comment by emillon** [comment]
Thanks! I rebased and tested that on the 5.0 branch.

File: src/dune_rules/lib.ml [hunk 7]
```diff
@@ -1207,6 +1224,16 @@ end = struct
...
     let open Memo.O in
     let open Resolved_deps_builder in
+    let deps =
+      let ocaml_version = db.lib_config.ocaml_version in
+      let bigarray_in_std_libraries =
+        Ocaml.Version.has_bigarray_library ocaml_version
+      in
+      if bigarray_in_std_libraries then deps
+      else
+        let bigarray = Lib_name.of_string "bigarray" in
```
Comment (rgrinberg) [inline src/dune_rules/lib.ml:1234]: This change is incorrect and introduces a bug that forbids local libraries named `bigarray` from being used on OCaml < 5.0.

Comment (moyodiallo) [reply src/dune_rules/lib.ml:1234]: True, I don't know how we can test if there a local library called `bigarray`.

Comment (emillon) [reply src/dune_rules/lib.ml:1234]: that has been fixed in #8902 .
